HTMLConverter should put accessibilityLabels into its NSAttributedString results

achristensen07 opened the pull request:
#### a37792319b43c53b0a231c49487c502245ec262d
<pre>
HTMLConverter should put accessibilityLabels into its NSAttributedString results
<a href="https://bugs.webkit.org/show_bug.cgi?id=250947">https://bugs.webkit.org/show_bug.cgi?id=250947</a>
rdar://103608678

Reviewed by Timothy Hatcher.

* Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::_addAttachmentForElement):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/259179@main">https://commits.webkit.org/259179@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f62623cd3b3a85858839d35d24917cb27570de4c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104245 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13334 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37157 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113456 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173746 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108177 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14411 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4247 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96469 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112508 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110014 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11093 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94160 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38757 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92953 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25771 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80422 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6695 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27131 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6829 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3685 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12844 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46684 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6321 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8615 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1316,6 +1317,12 @@ static Class _WebMessageDocumentClass()
         NSUInteger textLength = [_attrStr length];
         RetainPtr<NSTextAttachment> attachment = adoptNS([[PlatformNSTextAttachment alloc] initWithFileWrapper:fileWrapper.get()]);
+
+        if (auto& ariaLabel = element.getAttribute("aria-label"_s); !ariaLabel.isEmpty())
+            attachment.get().accessibilityLabel = ariaLabel;
+        if (auto& altText = element.getAttribute("alt"_s); !altText.isEmpty())
+            attachment.get().accessibilityLabel = altText;
```
Can we use `else if` here (and check for alt text first)?

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1316,6 +1317,12 @@ static Class _WebMessageDocumentClass()
+
+        if (auto& ariaLabel = element.getAttribute("aria-label"_s); !ariaLabel.isEmpty())
+            attachment.get().accessibilityLabel = ariaLabel;
+        if (auto& altText = element.getAttribute("alt"_s); !altText.isEmpty())
```
`altAttr` instead of `"alt"_s`.

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1316,6 +1317,12 @@ static Class _WebMessageDocumentClass()
         NSUInteger textLength = [_attrStr length];
         RetainPtr<NSTextAttachment> attachment = adoptNS([[PlatformNSTextAttachment alloc] initWithFileWrapper:fileWrapper.get()]);
+
+        if (auto& ariaLabel = element.getAttribute("aria-label"_s); !ariaLabel.isEmpty())
```
`aria_labelAttr` instead of `"aria-label"_s`.

pxlcoder commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -1316,6 +1317,12 @@ static Class _WebMessageDocumentClass()
         RetainPtr<NSTextAttachment> attachment = adoptNS([[PlatformNSTextAttachment alloc] initWithFileWrapper:fileWrapper.get()]);
+
+        if (auto& ariaLabel = element.getAttribute("aria-label"_s); !ariaLabel.isEmpty())
+            attachment.get().accessibilityLabel = ariaLabel;
```
Nit: We generally prefer to use square bracket syntax with `RetainPtr`.

`[attachment setAccessibilityLabel:ariaLabel]`

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
```diff
@@ -169,3 +169,37 @@
+        EXPECT_NOT_NULL(documentAttributes);
+        EXPECT_NULL(error);
+
+        __block bool foundImage1 { NO };
+        __block bool foundImage2 { NO };
```
`false` since this is a `bool`, as opposed to a `BOOL`.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
```diff
@@ -169,3 +169,37 @@
+        [attributedString enumerateAttribute:NSAttachmentAttributeName inRange:NSMakeRange(0, attributedString.length) options:0 usingBlock:^(id value, NSRange attributeRange, BOOL* stop) {
+            if ([value isKindOfClass:NSTextAttachment.class]) {
+                if ([[value accessibilityLabel] isEqualToString:@"alt text"])
+                    foundImage1 = YES;
+                if ([[value accessibilityLabel] isEqualToString:@"aria label text"])
+                    foundImage2 = YES;
```
`true` since the variables are `bool`s

achristensen07 committed:
HTMLConverter should put accessibilityLabels into its NSAttributedString results
https://bugs.webkit.org/show_bug.cgi?id=250947
rdar://103608678

Reviewed by Timothy Hatcher.

* Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::_addAttachmentForElement):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
(TEST):

Canonical link: https://commits.webkit.org/259179@main
